### DISCUSSION TITLE
Local hardware buttons (needs testing and webui work)

### DISCUSSION THREAD

**Discussion by o0shojo0o** [pixelit-project/PixelIt#111] [Ideas]
Continuing with Pullrequest #108:

Hi @miccgn, 
I finally got around to looking at this today and also "fixing" the WebUi. 
This all looks very good to me (great work), as I see it now only the implementation of the button action is missing right?

This is how the WebUi looks after the revision:
![image](https://user-images.githubusercontent.com/19333515/147882218-08e4ccd9-107d-46f1-a7e2-a407b78eb772.png)

**Comment by miccgn** [comment]
Hm, I just had a review of the master code, and couldn't find any issues. You should be able to select an action in that window (beneath Left Button action, where your cursor seems to be), and PixelIt should perform that very action. Does the action show up in the config file? ( /api/config )

**Reply by o0shojo0o** [reply]
Yes sorry, I have not seen yesterday the handling of the clock.
https://github.com/o0shojo0o/PixelIt/blob/2701d8de8089b876b216adb696d1c3977c47629a/src/PixelIt.ino#L824-L827

**Reply by miccgn** [reply]
Got it! First, I thought you would have missed the whole HandleAndSendButtonPress method, then I realized you already linted it.

**Comment by o0shojo0o** [comment]
Ok I will then today or tomorrow create a new release with the new features :)

**Comment by miccgn** [comment]
I can't tell about the MQTT features, but the buttons seem to work fine in my two PixelIts. Also the lux sensor and the BME680.

Maybe  I will do some work on DrawText soon, but don't wait for that.

**Reply by miccgn** [reply]
Can you put the new WebUI online? So I could give it a test run?

**Reply by o0shojo0o** [reply]
Yes, I can do that today around evening.
I wait then still with the release.

**Reply by miccgn** [reply]
There are two minor issues I yet couldn't get hold of:
- sometimes, sounds are not played, even if it is not a combination of volume and control frame. I.e. I am sending the very same frame for ten times, and nine out of these ten times, it plays.
- sometimes, scrolling text hangs for parts of a second. I experienced this some time ago with readings of the BME680. As a consequence, I introduced async reading. I need to check again if there's something else that could impact scrolling.

Both issues definitely are minor.

**Reply by o0shojo0o** [reply]
> Yes, I can do that today around evening.
> 
> I wait then still with the release.

Sorry, something came up today, but tomorrow I will upload the new version of the Ui.

**Reply by o0shojo0o** [reply]
From the WebUI is now in current build online, but I noticed this morning there is still an bug in the home area at the button view. 
But I did not have the time this morning to look at it more closely.

**Comment by miccgn** [comment]
> sometimes, sounds are not played, even if it is not a combination of volume and control frame. I.e. I am sending the very same frame for ten times, and nine out of these ten times, it plays.

I spent some time on this. Looks like a simply delay(100) might help. Another option seems to be to change the library to one that forces ack messages. However, after both attempts worked fine, the problem didn't occur with the base version any longer. So I will have to look into all alternatives some more. However, I won't have time for that on the weekend.

> sometimes, scrolling text hangs for parts of a second. I experienced this some time ago with readings of the BME680. As a consequence, I introduced async reading. I need to check again if there's something else that could impact scrolling.

Did not work on this yet.

**Reply by miccgn** [reply]
@o0shojo0o the issues seem to be minor, you may want to publish a relase - given that you are satisfied with the webui. Don't wait for my fixes to those issues.

**Reply by miccgn** [reply]
I just read https://github.com/PowerBroker2/DFPlayerMini_Fast/issues/26. It seems to be a real mess with different chips on DFPlayers...

Better don't see this as a PixelIt problem and go ahead with the release.

**Comment by poblabs** [comment]
I want to test this out too, but I am confused how it works. How would I setup left button to go to previous screen, right button to next screen and middle button to toggle stop animation on current screen or restart animation screens? I am using node-red for pixelit control. thanks!

**Reply by miccgn** [reply]
Actually, the device PixelIt itself is quite "stupid". It doesn't store previous or next screen. It only knows how to display the current screen. The order of screens to be shown (the playlist) is handled on node-red.
So, the pixelit controller on node-red would have to learn how to read the button presses, interpret them and send proper screens to the PixelIt device.

@o0shojo0o maybe you can say more?

**Reply by o0shojo0o** [reply]
Currently you would have to take the wen over MQTT to publish the key presses in Node-Red. 
As a workaround to jump one screen forward, one could send an empty alert scrren with 0 seconds display time. 
And workaround to end the animations you would have to store an empty playlist and when restarting the originals. 
At the moment there is nothing available for the buttons in the Node-Red PixelIt Node natively.

**Reply by o0shojo0o** [reply]
https://github.com/o0shojo0o/node-red-contrib-pixelit/issues/13
